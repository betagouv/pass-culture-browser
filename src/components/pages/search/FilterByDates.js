import get from 'lodash.get'
import moment from 'moment'
import PropTypes from 'prop-types'
import React, { PureComponent } from 'react'

import { DAYS_CHECKBOXES } from './utils'
import { DatePickerField } from '../../forms/inputs'

class FilterByDates extends PureComponent {
  constructor(props) {
    super(props)
    this.state = {
      pickedDate: null,
    }
    this.datepickerPopper = React.createRef()
  }

  componentDidUpdate() {
    const { initialDateParams, filterState } = this.props
    if (
      (initialDateParams && filterState.params.date === undefined) ||
      (initialDateParams && filterState.params.date === null)
    ) {
      this.setPickedDate(null)
    }
  }

  setPickedDate = value => {
    this.setState({
      pickedDate: value,
    })
  }

  onChange = day => {
    this.setPickedDate(null)
    const { filterActions, filterState } = this.props
    const pickedDaysInQuery = decodeURI(filterState.params.jours || '')
    const isdayAlreadyChecked = pickedDaysInQuery.includes(day)
    let callback
    const pickedDaysInQueryLenght = get(pickedDaysInQuery, 'length')

    // change callback value
    if (pickedDaysInQueryLenght === 0) {
      // Case 1
      const date = moment(moment.now()).toISOString()
      callback = () => filterActions.change({ date })
    } else if (
      isdayAlreadyChecked &&
      pickedDaysInQuery.split(',').length === 1
    ) {
      // Case 2
      callback = () => filterActions.change({ date: null })
    }

    if (isdayAlreadyChecked) {
      // Case 3 Callback is undefined
      filterActions.remove('jours', day, callback)
      return
    }
    // Case 4 add days
    filterActions.add('jours', day, callback)
  }

  onPickedDateChange = date => {
    const { filterActions } = this.props
    const formatedDate = (date && date.toISOString()) || null
    filterActions.change({ date: formatedDate, jours: null })
    this.setPickedDate(date)
  }

  isDaysChecked = (date, pickedDaysInQuery, inputValue) =>
    date !== null && inputValue === '0-1'
      ? false
      : pickedDaysInQuery.includes(inputValue)

  render() {
    const { pickedDate } = this.state
    const { filterState, minDate, title } = this.props
    const pickedDaysInQuery = decodeURI(filterState.params.jours || '')

    return (
      <div id="filter-by-dates" className="pt18">
        <h2 className="fs15 is-italic is-medium is-uppercase text-center mb12">
          {title}
        </h2>
        {/* FIXME: le scroll sous ios est pas terrible
          du fait que le input soit cliquable */}
        <div className="pc-scroll-horizontal is-relative pb18">
          <div className="pc-list flex-columns">
            {DAYS_CHECKBOXES.map(({ label, value }) => {
              const checked = this.isDaysChecked(
                pickedDate,
                pickedDaysInQuery,
                value
              )
              return (
                <label
                  key={value}
                  className="item flex-columns items-center py5 pl7 pr22"
                >
                  <span className="flex-0 field field-checkbox">
                    <input
                      type="checkbox"
                      className="input no-background"
                      checked={checked}
                      onChange={() => this.onChange(value)}
                    />
                  </span>
                  <span className="fs19 flex-1" style={{ whiteSpace: 'pre' }}>
                    {label}
                  </span>
                </label>
              )
            })}
            <DatePickerField
              name="pick-by-date-filter"
              className="item fs19 py5 px7"
              minDate={minDate}
              selected={pickedDate}
              onChange={this.onPickedDateChange}
              popperRefContainer={this.datepickerPopper}
            />
          </div>
        </div>
        <hr className="dotted-bottom-primary" />
        <div
          id="filter-by-dates-datepicker-popper-container"
          ref={this.datepickerPopper}
        />
      </div>
    )
  }
}

FilterByDates.defaultProps = {
  minDate: moment().format('MMM Do YY'),
}

FilterByDates.propTypes = {
  filterActions: PropTypes.object.isRequired,
  filterState: PropTypes.object.isRequired,
  initialDateParams: PropTypes.bool.isRequired,
  minDate: PropTypes.oneOfType([PropTypes.bool, PropTypes.string]),
  title: PropTypes.string.isRequired,
}
export default FilterByDates
