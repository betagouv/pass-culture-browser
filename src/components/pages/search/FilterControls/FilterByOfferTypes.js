import PropTypes from 'prop-types'
import React, { PureComponent } from 'react'

import SearchPicture from './SearchPicture'

class FilterByOfferTypes extends PureComponent {
  onChangeCategory = category => () => {
    const { filterActions, filterState } = this.props
    const typesValue = decodeURI(filterState.params.categories || '')
    const isAlreadyIncluded = typesValue.includes(category)

    if (isAlreadyIncluded) {
      filterActions.remove('categories', category)
      return
    }

    filterActions.add('categories', category)
  }

  render() {
    const { filterState, typeSublabels } = this.props
    const typesValue = decodeURI(filterState.params.categories || '')

    return (
      <div
        className="pt18"
        id="filter-by-offer-types"
      >
        <h2 className="fs15 is-italic is-medium is-uppercase text-center mb12">{'Quoi'}</h2>
        <div className="pc-scroll-horizontal is-relative">
          <div className="pc-list flex-columns pt7">
            {typeSublabels.map((category, index) => {
              const isChecked = typesValue.includes(category)
              const className = isChecked ? 'checked' : ''
              const inputId = `search-image-checkbox-${index}`

              return (
                <label
                  className={`item p3 is-relative ${className}`}
                  htmlFor={inputId}
                  key={category}
                >
                  <SearchPicture category={category} />
                  <input
                    checked={isChecked}
                    className="is-hidden"
                    id={inputId}
                    onChange={this.onChangeCategory(category)}
                    type="checkbox"
                    value={category}
                  />
                  {isChecked && (
                    <span className="icon-container is-absolute">
                      <span className="is-relative">
                        <i className="icon-legacy-check-circled" />
                      </span>
                    </span>
                  )}
                </label>
              )
            })}
          </div>
        </div>
      </div>
    )
  }
}

FilterByOfferTypes.propTypes = {
  filterActions: PropTypes.shape({
    add: PropTypes.func.isRequired,
    change: PropTypes.func.isRequired,
    remove: PropTypes.func.isRequired,
  }).isRequired,
  filterState: PropTypes.shape({
    isNew: PropTypes.bool,
    params: PropTypes.shape({
      categories: PropTypes.string,
      date: PropTypes.string,
      distance: PropTypes.string,
      jours: PropTypes.string,
      latitude: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
      longitude: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
      'mots-cles': PropTypes.string,
      orderBy: PropTypes.string,
    }),
  }).isRequired,
  typeSublabels: PropTypes.arrayOf(PropTypes.string).isRequired,
}

export default FilterByOfferTypes
