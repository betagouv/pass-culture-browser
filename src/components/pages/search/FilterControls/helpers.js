export const INFINITE_DISTANCE = 20000

export const distanceOptions = [
  {
    label: 'Toutes distances',
    value: INFINITE_DISTANCE,
  },
  {
    label: "Moins d'1 km",
    value: 1,
  },
  {
    label: 'Moins de 10 km',
    value: 10,
  },
  {
    label: 'Moins de 50 km',
    value: 50,
  },
]
