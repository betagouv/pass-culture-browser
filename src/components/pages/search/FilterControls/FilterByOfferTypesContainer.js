import { connect } from 'react-redux'

import FilterByOfferTypes from './FilterByOfferTypes'
import selectTypeSublabels from '../../../../selectors/data/typesSelector'

const mapStateToProps = state => ({
  typeSublabels: selectTypeSublabels(state),
})

export default connect(mapStateToProps)(FilterByOfferTypes)
