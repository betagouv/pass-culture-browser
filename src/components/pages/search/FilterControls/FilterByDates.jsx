import PropTypes from 'prop-types'
import React, { createRef, PureComponent } from 'react'

import DatePickerField from '../../../forms/inputs/DatePickerField/DatePickerField'
import { DAYS_CHECKBOXES, isDaysChecked } from '../helpers'

class FilterByDates extends PureComponent {
  constructor(props) {
    super(props)
    this.state = {
      pickedDate: null,
    }
    this.datepickerPopper = createRef()
  }

  componentDidUpdate() {
    const { filterState, shouldResetDate } = this.props

    if (shouldResetDate && filterState.params.date === undefined) {
      this.setPickedDate(null)
    }
  }

  setPickedDate = value => {
    this.setState({
      pickedDate: value,
    })
  }

  onChangeDate = day => () => {
    this.setPickedDate(null)
    const { filterActions, filterState } = this.props
    const pickedDaysInQuery = decodeURI(filterState.params.jours || '')
    const isDayAlreadyChecked = pickedDaysInQuery.includes(day)
    let callback

    if (pickedDaysInQuery.length === 0) {
      const date = new Date().toISOString()
      callback = () => filterActions.change({ date })
    } else if (isDayAlreadyChecked && pickedDaysInQuery.split(',').length === 1) {
      callback = () => filterActions.change({ date: null })
    }

    if (isDayAlreadyChecked) {
      filterActions.remove('jours', day, callback)
      return
    }

    filterActions.add('jours', day, callback)
  }

  handleOnChangePickedDate = (date = null) => {
    const { filterActions } = this.props

    const formattedDate =
      (date &&
        date
          .utcOffset(0)
          .add(100, 'milliseconds')
          .toISOString()) ||
      null
    filterActions.change({ date: formattedDate, jours: '0-0' })

    const pickedDate = (date && date.utcOffset(100)) || null
    this.setPickedDate(pickedDate)
  }

  render() {
    const { pickedDate } = this.state
    const { filterState } = this.props
    const pickedDaysInQuery = decodeURI(filterState.params.jours || '')

    return (
      <div id="filter-by-dates">
        <h2 className="fd-when-label">
          {'Quand'}
        </h2>
        <div className="fd-date-picker-container">
          <div className="fd-date-picker">
            <DatePickerField
              className="fd-date-picker-selectbox"
              minDate={new Date()}
              name="pick-by-date-filter"
              onChange={this.handleOnChangePickedDate}
              placeholder="Par date"
              selected={pickedDate}
            />
          </div>
        </div>
        <div className="fd-checkboxes">
          {DAYS_CHECKBOXES.map(({ label, value }, index) => {
            const checked = isDaysChecked(pickedDate, pickedDaysInQuery, value)

            return (
              <label
                className="flex-columns items-center py5 pl7"
                htmlFor={`filter-by-dates-days-${index}-checkbox`}
                key={value}
              >
                <input
                  checked={checked}
                  className="input form-checkbox field field-checkbox"
                  id={`filter-by-dates-days-${index}-checkbox`}
                  onChange={this.onChangeDate(value)}
                  type="checkbox"
                  value={value}
                />
                <span className="fs19">
                  {label}
                </span>
              </label>
            )
          })}
        </div>
        <hr className="dotted-bottom-primary" />
      </div>
    )
  }
}

FilterByDates.propTypes = {
  filterActions: PropTypes.shape({
    add: PropTypes.func.isRequired,
    change: PropTypes.func.isRequired,
    remove: PropTypes.func.isRequired,
  }).isRequired,
  filterState: PropTypes.shape({
    isNew: PropTypes.bool,
    params: PropTypes.shape({
      categories: PropTypes.string,
      date: PropTypes.string,
      distance: PropTypes.string,
      jours: PropTypes.string,
      latitude: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
      longitude: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
      'mots-cles': PropTypes.string,
      orderBy: PropTypes.string,
    }),
  }).isRequired,
  shouldResetDate: PropTypes.bool.isRequired,
}

export default FilterByDates
