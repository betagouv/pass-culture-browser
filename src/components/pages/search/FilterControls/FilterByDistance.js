import PropTypes from 'prop-types'
import React, { Fragment, PureComponent } from 'react'

import { distanceOptions, INFINITE_DISTANCE } from './helpers'

class FilterByDistance extends PureComponent {
  handleOnChangeDistance = event => {
    const { filterActions, geolocation } = this.props
    const distance = event.target.value
    let { latitude, longitude } = geolocation

    if (distance === INFINITE_DISTANCE) {
      latitude = null
      longitude = null
    }

    filterActions.change({ distance, latitude, longitude })
  }

  render() {
    const { filterState, geolocation } = this.props
    const geolocationActive = geolocation.latitude != null && geolocation.longitude != null
    const distanceValue = filterState.params.distance || INFINITE_DISTANCE

    return (
      <Fragment>
        <div
          className="distance-filter-options"
          id="filter-by-distance"
        >
          <h2 className="distance-filter-title">{'Où'}</h2>
          {!geolocationActive && (
            <div className="geoloc-warning">
              {'Activez votre géolocalisation pour utiliser ce filtre.'}
            </div>
          )}
          <select
            className="pc-selectbox pl24 py5 fs19"
            defaultValue={distanceValue}
            disabled={!geolocationActive}
            name="distance"
            onBlur={this.handleOnChangeDistance}
          >
            {distanceOptions.map(({ label, value }) => (
              <option
                key={value}
                value={value}
              >
                {label}
              </option>
            ))}
          </select>
        </div>
        <hr className="dotted-bottom-primary" />
      </Fragment>
    )
  }
}

FilterByDistance.propTypes = {
  filterActions: PropTypes.shape({
    add: PropTypes.func.isRequired,
    change: PropTypes.func.isRequired,
    remove: PropTypes.func.isRequired,
  }).isRequired,
  filterState: PropTypes.shape({
    isNew: PropTypes.bool,
    params: PropTypes.shape({
      categories: PropTypes.string,
      date: PropTypes.string,
      distance: PropTypes.string,
      jours: PropTypes.string,
      latitude: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
      longitude: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
      'mots-cles': PropTypes.string,
      orderBy: PropTypes.string,
    }),
  }).isRequired,
  geolocation: PropTypes.shape({
    latitude: PropTypes.number,
    longitude: PropTypes.number,
    watchId: PropTypes.number,
  }).isRequired,
}

export default FilterByDistance
