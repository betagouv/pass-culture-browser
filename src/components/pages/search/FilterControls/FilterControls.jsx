import PropTypes from 'prop-types'
import React, { Component } from 'react'
import { Transition } from 'react-transition-group'

import FilterByDates from './FilterByDates'
import FilterByDistanceContainer from './FilterByDistanceContainer'
import FilterByOfferTypesContainer from './FilterByOfferTypesContainer'
import { getCanFilter, getVisibleParamsAreEmpty, INITIAL_FILTER_PARAMS } from '../helpers'

const filtersPanelHeight = 500
const transitionDelay = 0
const transitionDuration = 500
const defaultStyle = {
  marginTop: `-${filtersPanelHeight}px`,
  transition: `margin-top ${transitionDuration}ms ease`,
}
const transitionStyles = {
  entered: { marginTop: 0 },
  entering: { marginTop: 0 },
  exited: { marginTop: `-${filtersPanelHeight}px` },
  exiting: { marginTop: `-${filtersPanelHeight}px` },
}

class FilterControls extends Component {
  constructor(props) {
    super(props)

    const { query } = props

    this.state = {
      shouldResetDate: true,
      params: query.parse() || {},
    }
  }

  componentDidUpdate(prevProps) {
    const { location } = this.props

    if (location.search !== prevProps.location.search) {
      this.handleReinitializeParams()
    }
  }

  handleReinitializeParams = () => {
    const { query } = this.props

    this.setState({
      shouldResetDate: true,
      params: query.parse(),
    })
  }

  handleOnClickFilterButton = () => {
    const { isVisible, onClickFilterButton, query } = this.props
    const { params } = this.state

    params.page = null
    query.change(params, { pathname: '/recherche/resultats' })
    this.setState({
      shouldResetDate: false,
    })
    onClickFilterButton(isVisible)
  }

  handleOnClickReset = () => {
    const { query } = this.props

    this.setState({
      shouldResetDate: true,
      params: {},
    })

    const nextQueryParams = { ...INITIAL_FILTER_PARAMS, page: null }
    query.change(nextQueryParams)
  }

  handleQueryChange = (newValue, callback) => {
    const { params } = this.state
    const nextFilterParams = { ...params, ...newValue }
    this.setState(
      {
        params: nextFilterParams,
      },
      callback
    )
  }

  handleQueryAdd = (paramKey, paramValue, callback) => {
    const { params } = this.state
    const encodedValue = encodeURI(paramValue)
    let nextValue = encodedValue
    const previousValue = params[paramKey]

    if (previousValue && previousValue.length) {
      nextValue = previousValue
        .split(',')
        .concat([encodedValue])
        .join(',')
    }

    this.handleQueryChange({ [paramKey]: nextValue }, callback)
  }

  handleQueryRemove = (paramKey, paramValue, callback) => {
    const { params } = this.state
    const previousValue = params[paramKey]

    if (previousValue && previousValue.length) {
      const encodedValue = encodeURI(paramValue)
      const nextValue = previousValue
        .split(',')
        .filter(value => value !== encodedValue)
        .join()

      this.handleQueryChange({ [paramKey]: nextValue === '' ? null : nextValue }, callback)
    }
  }

  render() {
    const { isVisible, query } = this.props
    const { shouldResetDate, params } = this.state
    const filterParamsAreEmpty = getVisibleParamsAreEmpty(params)
    const canFilter = getCanFilter(params, query.parse())

    const filterActions = {
      add: this.handleQueryAdd,
      change: this.handleQueryChange,
      remove: this.handleQueryRemove,
    }

    return (
      <div className="is-relative is-clipped">
        <Transition
          in={isVisible}
          timeout={transitionDelay}
        >
          {status => (
            <div
              className={`is-full-width transition-status-${status} mb20`}
              id="search-filter-menu"
              style={{ ...defaultStyle, ...transitionStyles[status] }}
            >
              <FilterByDates
                filterActions={filterActions}
                filterState={this.state}
                shouldResetDate={shouldResetDate}
              />
              <FilterByDistanceContainer
                filterActions={filterActions}
                filterState={this.state}
              />
              <FilterByOfferTypesContainer
                filterActions={filterActions}
                filterState={this.state}
              />
              <div
                className="flex-columns mt18"
                id="search-filter-menu-footer-controls"
              >
                <button
                  className="no-background no-outline col-1of2 fs20 py12"
                  disabled={filterParamsAreEmpty}
                  id="search-filter-reset-button"
                  onClick={this.handleOnClickReset}
                  type="reset"
                >
                  {'RÃ©initialiser'}
                </button>
                <button
                  className="no-background no-outline col-1of2 fs20 py12"
                  disabled={!canFilter}
                  id="filter-button"
                  onClick={this.handleOnClickFilterButton}
                  type="submit"
                >
                  <span className="is-bold">{'Filtrer'}</span>
                </button>
              </div>
            </div>
          )}
        </Transition>
      </div>
    )
  }
}

FilterControls.propTypes = {
  isVisible: PropTypes.bool.isRequired,
  location: PropTypes.shape().isRequired,
  onClickFilterButton: PropTypes.func.isRequired,
  query: PropTypes.shape().isRequired,
}

export default FilterControls
