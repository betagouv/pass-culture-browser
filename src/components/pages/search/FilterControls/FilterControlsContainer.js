import { connect } from 'react-redux'
import { compose } from 'redux'
import { deleteData } from 'redux-saga-data'

import FilterControls from './FilterControls'
import withFrenchQueryRouter from '../../../hocs/withFrenchQueryRouter'

const mapDispatchToProps = dispatch => ({
  resetSearchData: () => {
    dispatch(deleteData(null, { tags: ['search'] }))
  },
})

export default compose(
  withFrenchQueryRouter,
  connect(
    null,
    mapDispatchToProps
  )
)(FilterControls)
