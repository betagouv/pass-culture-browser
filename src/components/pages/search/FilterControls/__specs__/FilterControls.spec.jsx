import { mount, shallow } from 'enzyme'
import { createBrowserHistory } from 'history'
import React from 'react'
import { Provider } from 'react-redux'
import { Route, Router } from 'react-router-dom'
import { Transition } from 'react-transition-group'
import configureStore from 'redux-mock-store'

import FilterByDates from '../FilterByDates'
import FilterByDistanceContainer from '../FilterByDistanceContainer'
import FilterByOfferTypesContainer from '../FilterByOfferTypesContainer'
import FilterControls from '../FilterControls'
import FilterControlsContainer from '../FilterControlsContainer'
import { INITIAL_FILTER_PARAMS } from '../../helpers'
import REDUX_STATE from '../../../../../mocks/reduxState'

describe('src | components | pages | search | FilterControls', () => {
  let props

  beforeEach(() => {
    props = {
      isVisible: true,
      location: {},
      onClickFilterButton: jest.fn(),
      query: {
        change: jest.fn(),
        parse: jest.fn(),
      },
      resetSearchStore: jest.fn(),
    }
  })

  it('should match the snapshot', () => {
    // when
    const wrapper = shallow(<FilterControls {...props} />)

    // then
    expect(wrapper).toMatchSnapshot()
  })

  describe('constructor()', () => {
    it('should initialize state with current query arguments', () => {
      // given
      const queryParams = { prop: 'mocked object' }
      jest.spyOn(props.query, 'parse').mockImplementation(() => queryParams)
      const wrapper = shallow(<FilterControls {...props} />)

      // when
      const { params } = wrapper.instance().state

      // then
      expect(params).toStrictEqual(queryParams)
    })
  })

  describe('onComponentDidUpdate', () => {
    it('should reinitialize the state if the URL has changed', () => {
      // given
      const mockOnClickFilterButton = jest.fn()
      const store = configureStore([])(REDUX_STATE)
      const history = createBrowserHistory()
      history.push('/test?categories=Jouer')
      const wrapper = mount(
        <Provider store={store}>
          <Router history={history}>
            <Route path="/test">
              <FilterControlsContainer
                isVisible
                onClickFilterButton={mockOnClickFilterButton}
              />
            </Route>
          </Router>
        </Provider>
      )

      // when
      history.push('/test?categories=Sourire')

      // then
      const searchFilter = wrapper.find('FilterControls')
      expect(searchFilter.state('filterParamsMatchingQueryParams')).toBe(false)
      expect(searchFilter.state('initialDateParams')).toBe(true)
      expect(searchFilter.state('params')).toStrictEqual(expect.any(Object))
    })
  })

  describe('handleOnClickFilterButton()', () => {
    it('should close the filters and change the URL by default', () => {
      // given
      const wrapper = shallow(<FilterControls {...props} />)

      // when
      wrapper.instance().handleOnClickFilterButton()

      // then
      expect(props.onClickFilterButton).toHaveBeenCalledWith(props.isVisible)
    })

    it('should reset the store when filter params match the query params', () => {
      // given
      const wrapper = shallow(<FilterControls {...props} />)
      wrapper.setState({ filterParamsMatchingQueryParams: true })

      // when
      wrapper.instance().handleOnClickFilterButton()

      // then
      expect(props.resetSearchStore).toHaveBeenCalledWith()
    })
  })

  describe('handleOnClickReset()', () => {
    it('should initialize default filter params and reset the URL', () => {
      // given
      const wrapper = shallow(<FilterControls {...props} />)

      // when
      wrapper.instance().handleOnClickReset()

      // then
      const expected = {
        filterParamsMatchingQueryParams: false,
        initialDateParams: true,
        params: {},
      }
      expect(props.resetSearchStore).toHaveBeenCalledWith()
      expect(wrapper.state()).toStrictEqual(expected)
      expect(props.query.change).toHaveBeenCalledWith(INITIAL_FILTER_PARAMS, {
        pathname: '/recherche/resultats/tout',
      })
    })
  })

  describe('handleQueryAdd()', () => {
    it('should add filter parameters when I have not initial parameters', () => {
      // given
      props.query.parse = () => ({})
      const wrapper = shallow(<FilterControls {...props} />)

      // when
      wrapper.instance().handleQueryAdd('jours', '0-1', jest.fn())

      // then
      expect(wrapper.state()).toStrictEqual({
        filterParamsMatchingQueryParams: 'jours',
        initialDateParams: true,
        params: {
          jours: '0-1',
        },
      })
    })

    it('should add filter parameters when I have initial parameters', () => {
      // given
      props.query.parse = () => ({
        categories: 'Jouer',
      })
      const wrapper = shallow(<FilterControls {...props} />)

      // when
      wrapper.instance().handleQueryAdd('categories', 'Lire', undefined)

      // then
      expect(wrapper.state()).toStrictEqual({
        filterParamsMatchingQueryParams: 'categories',
        initialDateParams: true,
        params: {
          categories: 'Jouer,Lire',
        },
      })
    })
  })

  describe('handleQueryRemove()', () => {
    it('should remove filter parameters when I have initial parameters', () => {
      // given
      props.query.parse = () => ({
        categories: 'Jouer,Applaudir',
      })
      const wrapper = shallow(<FilterControls {...props} />)

      // when
      wrapper.instance().handleQueryRemove('categories', 'Jouer', undefined)

      // then
      expect(wrapper.state()).toStrictEqual({
        filterParamsMatchingQueryParams: 'categories',
        initialDateParams: true,
        params: {
          categories: 'Applaudir',
        },
      })
    })
  })

  describe('render()', () => {
    it('should have three filters and one reset button and one filter button by default', () => {
      // given
      const store = configureStore([])(REDUX_STATE)
      const wrapper = mount(
        <Provider store={store}>
          <FilterControls {...props} />
        </Provider>
      )

      // when
      const transition = wrapper.find(Transition)
      const filterByDates = transition.find(FilterByDates)
      const filterByDistanceContainer = transition.find(FilterByDistanceContainer)
      const filterByOfferTypesContainer = transition.find(FilterByOfferTypesContainer)
      const resetButton = transition.find('#search-filter-reset-button')
      const filterButton = transition.find('#filter-button')

      // then
      expect(transition).toHaveLength(1)
      expect(filterByDates).toHaveLength(1)
      expect(filterByDistanceContainer).toHaveLength(1)
      expect(filterByOfferTypesContainer).toHaveLength(1)
      expect(resetButton).toHaveLength(1)
      expect(filterButton).toHaveLength(1)
    })
  })
})
