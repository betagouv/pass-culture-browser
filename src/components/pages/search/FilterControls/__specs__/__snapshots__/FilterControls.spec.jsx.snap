// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`src | components | pages | search | FilterControls should match the snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <FilterControls
    isVisible={true}
    location={Object {}}
    onClickFilterButton={[MockFunction]}
    query={
      Object {
        "change": [MockFunction],
        "parse": [MockFunction] {
          "calls": Array [
            Array [],
          ],
          "results": Array [
            Object {
              "type": "return",
              "value": undefined,
            },
          ],
        },
      }
    }
    resetSearchData={[MockFunction]}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": <Transition
        appear={false}
        enter={true}
        exit={true}
        in={true}
        mountOnEnter={false}
        onEnter={[Function]}
        onEntered={[Function]}
        onEntering={[Function]}
        onExit={[Function]}
        onExited={[Function]}
        onExiting={[Function]}
        timeout={0}
        unmountOnExit={false}
      >
        [Function]
      </Transition>,
      "className": "is-relative is-clipped",
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "appear": false,
        "children": [Function],
        "enter": true,
        "exit": true,
        "in": true,
        "mountOnEnter": false,
        "onEnter": [Function],
        "onEntered": [Function],
        "onEntering": [Function],
        "onExit": [Function],
        "onExited": [Function],
        "onExiting": [Function],
        "timeout": 0,
        "unmountOnExit": false,
      },
      "ref": null,
      "rendered": [Function],
      "type": [Function],
    },
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": <Transition
          appear={false}
          enter={true}
          exit={true}
          in={true}
          mountOnEnter={false}
          onEnter={[Function]}
          onEntered={[Function]}
          onEntering={[Function]}
          onExit={[Function]}
          onExited={[Function]}
          onExiting={[Function]}
          timeout={0}
          unmountOnExit={false}
        >
          [Function]
        </Transition>,
        "className": "is-relative is-clipped",
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "appear": false,
          "children": [Function],
          "enter": true,
          "exit": true,
          "in": true,
          "mountOnEnter": false,
          "onEnter": [Function],
          "onEntered": [Function],
          "onEntering": [Function],
          "onExit": [Function],
          "onExited": [Function],
          "onExiting": [Function],
          "timeout": 0,
          "unmountOnExit": false,
        },
        "ref": null,
        "rendered": [Function],
        "type": [Function],
      },
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromError": true,
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    Symbol(enzyme.__providerValues__): undefined,
  },
  Symbol(enzyme.__providerValues__): Map {},
  Symbol(enzyme.__childContext__): null,
}
`;
